hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/MasterQuotasObserver.java

```diff
@@ -78,6 +79,20 @@ public void postDeleteTable(
           admin.setQuota(settings);
         }
       }
+    } else if (quotasAtNamespace != null) {
+      // If quota present at namespace level remove the table entry from 'hbase:quota'
```

Comment:
Why is this in `postDeleteTable`? The quota should be dropped if the namespace is dropped, which should happen in `postDeleteNamespace`, right?

Reply:
Because even though table is dropped it's entry is still present in the 'hbase:quota' table because of the namespace quota. If we create another table with same name inside that namespace it inherits the earlier quota settings. So during dropping the table we need to remove that entry/setting from the hbase:quota for our table even though quota was present at namespace level.

Reply:
> Because even though table is dropped it's entry is still present in the 'hbase:quota' table because of the namespace quota. If we create another table with same name inside that namespace it inherits the earlier quota settings.

That's by design, not a bug. If you set a quota on a namespace (as opposed to a quota on the table), you would automatically get that quota for all tables in that namespace.

This allows an organization to be given free-rule over some namespace while still enforcing an upper-bound on the allowed resources in HBase. e.g. a namespace space quota of 200G would allow 200G of data in any number of tables in that namespace, not mattering which table uses that quota.

Now, if you drop a namespace and the quota still remains, that's a bug, but I don't think that's what you're trying to solve.

Reply:
Hi Josh, thanks for the review , I have one doubt.

1. create namespace with quota.
2. create table.
3. violate quota on the namespace by puting data into the table.
4. delete the table.
5. create another table with same name inside namespace.
6. Currently , table will be in violation even though it doesn't have any data in it since it's entry was not deleted from 'hbase:quota'.

Is this the intended behaviour?

Reply:
> Currently , table will be in violation even though it doesn't have any data in it since it's entry was not deleted from 'hbase:quota'.

So let's say you have `t1` and `t1'` which are the two instances of the tables as you describe.

If I had to guess, when you create `t1'`, the system would still see the `t1`'s region size reports for a period of time (until `t1'` reports its for the first time).

If this is indeed what's happening, the fix would be to purge the region size reports for a table when it's deleted, not try to alter the state of `hbase:quota`.  Region size reports are held in memory inside of the active HBase Master inside of the class `MasterQuotaManager`. There _is no_ quota defined at the table level, so, again, I'm not sure what you expect this change to be doing directly. If your test works, I can only imagine that it's by circumstance.